tests/__init__.py

```diff
@@ -359,3 +362,66 @@ def assert_url_equal(url1, url2):
     assert_equal(parts1.hostname, parts2.hostname)
     assert_equal(parts1.port, parts2.port)
     assert_equal(parse_qs(parts1.query), parse_qs(parts2.query))
+
+
+class RawResponse(BytesIO):
+    def stream(self, **kwargs):
+        contents = self.read()
+        while contents:
+            yield contents
+            contents = self.read()
+
+
+class BotocoreHTTPStubber(object):
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self._requests = []
+        self._responses = []
+
+    def add_response(self, response):
+        self._responses.append(response)
+
+    def create_response(self, url=None, status=200, headers=None, body=None):
+        if url is None:
+            url = 'https://example.com'
+
+        if headers is None:
+            headers = {}
+
+        if body is None:
+            body = b''
+
+        raw = RawResponse(body)
+        response = AWSResponse(url, status, headers, raw)
+        self._responses.append(response)
+
+    @property
+    def requests(self):
+        return self._requests
+
+    @property
+    def request_count(self):
+        return len(self._requests)
+
+    @contextlib.contextmanager
+    def wrap_client(self, client):
+        events = client.meta.events
+        events.register('send-request', self)
+        try:
+            yield client
+        finally:
+            events.unregister('send-request', self)
+
+    def __call__(self, **kwargs):
+        request = kwargs.get('request')
+        self._requests.append(request)
+        if self._responses:
+            response = self._responses.pop(0)
+            if isinstance(response, Exception):
```

Comment:
Internally, it may be worth exposing the items of the list as a dictionary/namedtuple with the following format (or something like this):

```
{
    "type": "<error|response>",
    "payload": response
}
```
So instead of having to check what the instance is, you can just look at the `type`. Furthermore, it should be pretty easy to do this if we had explicit methods for adding responses and errors.

Reply:
I am fine with how you have it now. This comment was more related to how to generalize it a bit more if there was an `add_error` to the stubber.